### PR TITLE
Add Basic Structure

### PR DESCRIPTION
**Pull request by sharanry**
#1 

This adds some basic types and functions for `LatentGP`s. Currently, I have added only Gaussian likelihood as a sanity test. I will be adding more likelihoods soon based on your suggestions either in this PR or a separate one.

@willtebbutt

### CODE REVIEW

File: src/latent_gp.jl [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+ - `f` is a Gaussian Process.
+ - `x` is an `AbstractVector` of inputs of type `Tx`.
+ - `Φ` is the log likelihood function which maps input of type `Tx` to `Real`.
+    
+"""
+struct LatentGP{T<:GP,Tx}
+    f::T
+    x::AbstractVector{Tx}
+    σ²
```
Comment (willtebbutt) [inline src/latent_gp.jl:12]: I'm don't think that we need a noise variance parameter for this model, do we? This should be handled by the likelihood now.

edit: this is superseded by my next comment. In general you'll want the noise variance in `fx` to be something close to zero, but typically it won't quite be zero for stability reasons. Either way, we shouldn't have to worry about this too much in this package I don't think, at least not at the API level.

File: src/latent_gp.jl [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    end
+
+
+function mean(fx::LatentGP)
+    return AbstractGPs.mean(fx.f, fx.x)
+end
+
+function cov(fx::LatentGP)
+    return AbstractGPs.cov(fx.f, fx.x)
+end
```
Comment (willtebbutt) [inline src/latent_gp.jl:30]: I don't think that it makes sense to provide these functions in this package. I think we need to discuss this a bit further in #1 to make sure that we're on the same page here.

File: src/latent_gp.jl (hunk 1 above)
Comment (willtebbutt) [reply src/latent_gp.jl:12]: Also, I think that it makes sense to just keep a single `FiniteGP` object `fx` rather than separately storing `f` and `x`.

Comment (sharanry) [reply src/latent_gp.jl:12]: If we were to keep a single `FiniteGP` object, it would incorporate the noise too. This should solve both the issues.

File: src/latent_gp.jl (hunk 2 above)
Comment (sharanry) [reply src/latent_gp.jl:30]: If we were to maintain `Distributions.jl` like API, we should probably keep these?

Comment (willtebbutt) [reply src/latent_gp.jl:30]: Ideally we would do this, but it's not generally going to be clear how to compute either the mean or covariance over the output (in the proposal 2 sense). For example if the likelihood is some horrible non-linear thing.

File: src/latent_gp.jl [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ - `Φ` is the log likelihood function which maps input of type `Tx` to `Real`.
+    
+"""
+struct LatentGP{T<:AbstractGPs.FiniteGP}
+    fx::T
+    Φ
+    function LatentGP(fx::AbstractGPs.FiniteGP, Φ) where Tx
+        return new{typeof(fx)}(fx, Φ)
+    end    
+end
```
Comment (willtebbutt) [inline src/latent_gp.jl:14]:
```suggestion
struct LatentGP{Tfx<:AbstractGPs.FiniteGP, TΦ}
    fx::Tfx
    Φ::TΦ
end
```

File: src/latent_gp.jl [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+function LatentGP(f::GP, x::AbstractVector{Tx}, σ², Φ) where Tx
+        return LatentGP(f(x, σ²), Φ)
+end
+
+function LatentGP(f::GP, x::AbstractVector{Tx}, Φ) where Tx
+        return LatentGP(f(x), Φ)
+end
```
Comment (willtebbutt) [inline src/latent_gp.jl:22]: Do we really need these additional methods? Is it not sufficient to have
```julia
LatentGP(f(x, σ²), Φ)
```
as the unique + correct way to construct a `LatentGP`?

File: src/likelihoods.jl [hunk 5]
```diff
@@ -0,0 +1,10 @@
+
+"""
+    gaussian(fx::LatentGP, y)
+
+Returns gaussian log-likelihood of output `y` w.r.t to the given `LatentGP`.  
+"""
+function gaussian(fx::LatentGP, y)
+    return Distributions.logpdf(MvNormal(mean(fx), cov(fx) + fx.σ² * I), y)
+end
+
```
Comment (willtebbutt) [inline src/likelihoods.jl:10]: Under the new design, do we need this?

Comment (sharanry) [reply src/likelihoods.jl:10]: Nope. I have more on this for you on slack.

File: src/latent_gp.jl (hunk 3 above)
Comment (sharanry) [reply src/latent_gp.jl:14]: Thanks. Didn't notice it to be the default constructor. :sweat_smile:

File: src/latent_gp.jl (hunk 4 above)
Comment (sharanry) [reply src/latent_gp.jl:22]: Nope. Now that you point it out, they seem unnecessary.

File: Project.toml [hunk 6]
```diff
@@ -3,8 +3,19 @@ uuid = "45b822f7-b0f9-4a40-9473-dffa0412f8e7"
...
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
+
 [compat]
-julia = "1"
+AbstractGPs = "0.1.1"
+Distributions = "0.19, 0.20, 0.21, 0.22, 0.23"
+KernelFunctions = "0.4"
+julia = ">=1.3"
```
Comment (willtebbutt) [inline Project.toml:18]:
```suggestion
julia = "1.3"
```
This means the same thing :)

**Comment by theogf** [comment]
Two general comments :
- I think it would make it generally easier to keep all likelihoods separated (one folder likelihoods and one likelihood per file)
- For likelihoods with a transformation on `f` like Bernoulli or Poisson, it would be nicer to create an additional `link` (like a type) to leave the transformation choice free. (One can then choose between `logit` and `probit` for Bernoulli, `exp`, `square` or `logistic` for Poisson etc

**Comment by willtebbutt** [comment]
> I think it would make it generally easier to keep all likelihoods separated (one folder likelihoods and one likelihood per file)

This seems reasonable to me.

> For likelihoods with a transformation on f like Bernoulli or Poisson, it would be nicer to create an additional link (like a type) to leave the transformation choice free. (One can then choose between logit and probit for Bernoulli, exp, square or logistic for Poisson etc

Agreed that the user should be able to choose the transformation. I don't think that we need to constrain it to be a particular type though -- it's sufficient to let the user just provide a function of their choosing imho.

File: src/likelihoods.jl [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+```math
+    p(y|f) = Normal(y | f, σ²)
+```
+On calling, this would return a normal distribution with mean `f` and variance σ².
+"""
+struct GaussianLikelihood{T<:Real} <: Likelihood
+    σ²::Vector{T}
+    function GaussianLikelihood(σ²::T) where {T<:Real}
+        new{typeof(σ²)}([σ²])
+    end
```
Comment (willtebbutt) [inline src/likelihoods.jl:24]:
```suggestion
```

I don't think that we need this.

Comment (sharanry) [reply src/likelihoods.jl:24]: Flux's `params()` seems to fail if the fields are not a vector. I found this kind of a [work around](https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/blob/master/src/basekernels/constant.jl#L51-L56) in `KernelFunctions.jl`

Comment (willtebbutt) [reply src/likelihoods.jl:24]: Oh, sorry, yeah, you're totally right.

Comment (theogf) [reply src/likelihoods.jl:24]: That would be amazing if one could use `Ref`... Do we know if there is any plan on this?

File: src/trainable.jl [hunk 8]
```diff
@@ -0,0 +1,5 @@
+import .Flux.trainable
+
+Flux.trainable(::Likelihood) = ()
```
Comment (theogf) [inline src/trainable.jl:3]: This is not needed, this is already the default behavior

File: src/latent_gp.jl [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+struct LatentGP{T<:AbstractGPs.FiniteGP}
+    fx::T
+    Φ
+end
+
+function Distributions.rand(rng::AbstractRNG, lgp::LatentGP)
+    f = rand(rng, lgp.fx)
+    y = rand(rng, lgp.Φ(f))
+    return (f=f, y=y)
+end
```
Comment (theogf) [inline src/latent_gp.jl:17]: I think this would be nice to create `rand` functions also to just get `f` or just get `y`.
Like
```
randf (rng::AbstractRNG, lgp::LatentGP) = rand(rng, lgp.fx)
randy(rng::AbstractRNG, lgp::LatentGP) = rand(rng, lgp).y
```

**Comment by theogf** [comment]
> Agreed that the user should be able to choose the transformation. I don't think that we need to constrain it to be a particular type though -- it's sufficient to let the user just provide a function of their choosing imho.

Well some links can contain hyperparameters, hence the need for a `struct` (a bit in the same way we have `Transform` in `KernelFunctions.jl`.
For example for Poisson one parametrization is `C * logistic(f)` where `C` needs to be optimized.

File: src/latent_gp.jl [hunk 10]
```diff
@@ -0,0 +1,29 @@
+"""
+    LatentGP(f<:GP, x<:AbstractVector, Φ)
+
+ - `fx` is a `FiniteGP`.
+ - `Φ` is the log likelihood function which maps input of type `Tx` to `Real`.
+    
+"""
+struct LatentGP{T<:AbstractGPs.FiniteGP}
+    fx::T
+    Φ
```
Comment (devmotion) [inline src/latent_gp.jl:10]:
```suggestion
struct LatentGP{T<:AbstractGPs.FiniteGP,S}
    fx::T
    Φ::S
```
Maybe also don't use unicode characters here since it's not just an internal implementation detail but users might want to access the field.

File: src/latent_gp.jl (hunk 9 above)
Comment (willtebbutt) [reply src/latent_gp.jl:17]: I disagree. What's wrong with
```julia
rand(rng, lgp).f
rand(rng, lgp).y
```
?

Comment (theogf) [reply src/latent_gp.jl:17]: If you are only interested in `f` you allocate `y` for nothing. But I agree that `randy` is not needed

Comment (willtebbutt) [reply src/latent_gp.jl:17]: Sure. If you're worried about performance you can just call
```julia
rand(rng, lgp.fx)
```

Comment (willtebbutt) [reply src/latent_gp.jl:17]: This is not resolved, I really can't see the advantage of having `randf` and `randy`.

**Comment by willtebbutt** [comment]
> Well some links can contain hyperparameters, hence the need for a struct (a bit in the same way we have Transform in KernelFunctions.jl.
For example for Poisson one parametrization is C * logistic(f) where C needs to be optimized.

Sure -- and if the user wants to parametrise it that way that's fine. I'm just saying that we shouldn't require that the linking operation subtypes a particular thing.

File: src/latent_gp.jl (hunk 9 above)
Comment (theogf) [reply src/latent_gp.jl]: I agree we can forget `randy`, but I would still insist that `randf` can be a useful wrapper.

Comment (willtebbutt) [reply src/latent_gp.jl]: But why?

**Comment by theogf** [comment]
> Sure -- and if the user wants to parametrise it that way that's fine. I'm just saying that we shouldn't require that the linking operation subtypes a particular thing.

Sounds good. Just wanted to make sure all information was there :grin:

**Comment by sharanry** [comment]
@theogf Do you have any specific "links" in mind? A few basic ones which come to my mind are:
- Scalar
- Logistic
- Softmax
- Compostion


@willtebbutt A random thought. Do you think we can use Bijectors.jl for links? We probably just need a subset of its capabilities.

Also, IMO this should be a separate PR.

File: src/latent_gp.jl (hunk 9 above)
Comment (sharanry) [reply src/latent_gp.jl]: After a second look, I too feel it is better not have `randf` or `randy`. All the components required are exposed and (if required) we could mention this usage in the docstring.

**Comment by willtebbutt** [comment]
> @willtebbutt A random thought. Do you think we can use Bijectors.jl for links? We probably just need a subset of its capabilities.

This sounds reasonable, but I agree that it's something for another PR. In any case, it's also the kind of thing that we can include as a "sensible default", but not something that needs to be enforced.

File: src/latent_gp.jl [hunk 11]
```diff
@@ -0,0 +1,29 @@
+"""
+    LatentGP(f<:GP, x<:AbstractVector, lik)
+
+ - `fx` is a `FiniteGP`.
+ - `lik` is the log likelihood function which maps input of type `Tx` to `Real`.
+
```
Comment (devmotion) [inline src/latent_gp.jl:6]: The docstring is incorrect (type signature) and the formatting should be fixed.

File: src/likelihoods/likelihoods.jl [hunk 12]
```diff
@@ -0,0 +1,10 @@
+"""
+    Likelihood
+
+An abstract type for likelihoods. Likelihoods are used to model the uncertainities associated
+with the data.
+"""
+abstract type Likelihood end
```
Comment (devmotion) [inline src/likelihoods/likelihoods.jl:7]: Is the abstract type needed and part of the design? I'm not sure if it's actually useful since no methods are defined for it and it's not part of any type restrictions (and I guess we don't want these restrictions to allow for any kinds of likelihoods).

File: src/LatentGPs.jl [hunk 13]
```diff
@@ -1,5 +1,28 @@
...
+
+export GaussianLikelihood, PoissonLikelihood
+
+
+include("latent_gp.jl")
+include("likelihoods/likelihoods.jl")
+
+function __init__()
+    @require Flux="587475ba-b771-5e3f-ad9e-33799f191a9c" include("trainable.jl")
+end
```
Comment (devmotion) [inline src/LatentGPs.jl:25]: I don't think this is needed anymore (not in KernelFunctions either). The logic was separated into https://github.com/FluxML/Functors.jl. I.e., I guess it would be sufficient to depend on Functors (which should be a light dependency we can take I guess?), and just add
```julia
@functor GaussianLikelihood
```
after the definition of `GaussianLikelihood`.

Comment (devmotion) [reply src/LatentGPs.jl:25]: Would be nice to get rid of Requires since it can increase loading times quite a bit, and it would be more generally applicable and not only be restricted to training with Flux.

File: src/likelihoods/likelihoods.jl (hunk 12 above)
Comment (sharanry) [reply src/likelihoods/likelihoods.jl:7]: It might be useful while defining links which @theogf suggested earlier in [this comment](https://github.com/JuliaGaussianProcesses/LatentGPs.jl/pull/3#issuecomment-634599211).

Comment (willtebbutt) [reply src/likelihoods/likelihoods.jl:7]: How will it help when defining links?

Comment (sharanry) [reply src/likelihoods/likelihoods.jl:7]: This depends on how we design Links. One possible way while making use of `Bijectors.jl` would be
```julia
struct Link
    lik::Likelihood
    transform::Bijector
end
(link::Link)(x) = link.lik(link.transform(x))

#while making a LatentGP
link = Link(PoissonLikelihood(), Exp())
lgp = LatentGP(gp, link)
```
We could maybe come up with something more self-explanatory than "Link".

Comment (devmotion) [reply src/likelihoods/likelihoods.jl:7]: But that's not actually an argument in favour of a `Likelihood` type, you probably just want to have
```julia
struct Link{L,B<:Bijector}
    lik::L
    transform::B
end
```
to allow all kinds of user-defined likelihood functions. `Likelihood` is not used anywhere else in the example, it seems.

Comment (sharanry) [reply src/likelihoods/likelihoods.jl:7]: Yeah you are right. I will remove Likelihood type. Thanks!

**Comment by willtebbutt** [comment]
Sorry for letting this drop off the radar a bit. @sharanry @devmotion @theogf is there anything else in particular that any of you want to get done in this PR? I'm happy for it to be merged I think.

File: src/LatentGPs.jl [hunk 14]
```diff
@@ -1,5 +1,26 @@
...
+using LinearAlgebra
+using Random
+using Functors
+
+import Statistics
+import Distributions
+
+export LatentGP
+
+export logpdf, rand, trainable, randf, randy
```
Comment (devmotion) [inline src/LatentGPs.jl:15]: It seems currently only
```suggestion
export logpdf, rand
```
are defined.

File: src/likelihoods/gaussian.jl [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+"""
+struct GaussianLikelihood{T<:Real}
+    σ²::T
+end
+
+GaussianLikelihood() = GaussianLikelihood(1e-6)
+
+@functor GaussianLikelihood
+
+(l::GaussianLikelihood)(f::Real) = Normal(f, l.σ²)
```
Comment (devmotion) [inline src/likelihoods/gaussian.jl:20]: This should be
```suggestion
(l::GaussianLikelihood)(f::Real) = Normal(f, sqrt(l.σ²))
```

File: src/likelihoods/gaussian.jl [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+(l::GaussianLikelihood)(fs::AbstractVector{<:Real}) = Product([Normal(f, l.σ²) for f in fs])
```
Comment (devmotion) [inline src/likelihoods/gaussian.jl:22]: Maybe it might make sense to avoid the iteration and allocation of the intermediate array of distributions and just define
```suggestion
(l::GaussianLikelihood)(fs::AbstractVector{<:Real}) = MvNormal(fs, sqrt(l.σ²))
```

In general I guess one should avoid calling `Product` directly but instead use `product` (same as `Truncated` vs `truncated`) since `product` allows to define (potentially) more efficient dispatches. For instance, here `product` would create just a `MvNormal` distribution.

**Comment by devmotion** [comment]
Maybe one could configure CompatHelper and use a `test/Project.toml` setup right away.

File: src/likelihoods/gaussian.jl (hunk 16 above)
Comment (sharanry) [reply src/likelihoods/gaussian.jl]: I wasn't aware of the difference between `Product` and `product`. Thank you!

Comment (devmotion) [reply src/likelihoods/gaussian.jl]: Oh I just remembered that it's actually called `product_distribution` (I guess `product` would be a too generic name :smile:).

**Comment by sharanry** [comment]
Does anyone have any objections to merge this?

**Comment by willtebbutt** [comment]
Nope, but I think we'll need to do some refactoring based on #1

**Comment by sharanry** [comment]
> Nope, but I think we'll need to do some refactoring based on #1

Can we do this as a separate PR?

**Comment by willtebbutt** [comment]
Yes, definitely. Please go ahead and merge this :)
